File: javascript/json_schema_validator.js
#!/usr/bin/env node
// json_schema_validator.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/json_schema_validator.py
// auch in: OpenClaw@gateway2:skills/json-utils/scripts/json_schema_validator.py
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

/**
 * JSON Schema Validator - Validiert JSON gegen JSON Schema Draft 7/2020-12.
 * Erweitert Pydantic mit externen Schema-Dateien.
 */

const fs = require('fs');
const path = require('path');

// Dynamically import modules that might not be available
let HAS_JSONSCHEMA = false;
let validateFunction = null;
let JSONSchemaValidationError = null;

let HAS_AJV = false;
let Ajv = null;
let ajvInstance = null;

// Try to load jsonschema
try {
  const jsonschema = require('jsonschema');
  validateFunction = jsonschema.validate;
  JSONSchemaValidationError = jsonschema.ValidationError;
  HAS_JSONSCHEMA = true;
} catch (e) {
  // jsonschema not available
}

// Try to load ajv as fallback
try {
  Ajv = require('ajv');
  const addFormats = require('ajv-formats');
  ajvInstance = new Ajv({allErrors: true});
  addFormats(ajvInstance);
  HAS_AJV = true;
} catch (e) {
  // ajv not available
}

// Custom error classes
class JSONValidationError extends Error {
  constructor(message) {
    super(message);
    this.name = 'JSONValidationError';
  }
}

class SchemaValidationError extends JSONValidationError {
  /** Raised when JSON Schema validation fails. */
  constructor(message) {
    super(message);
    this.name = 'SchemaValidationError';
  }
}

class JSONProcessingError extends Error {
  constructor(message) {
    super(message);
    this.name = 'JSONProcessingError';
  }
}

/**
 * Parst JSON mit optionaler Reparatur.
 * 
 * @param {string} raw_input - JSON-String
 * @param {boolean} repair - Ob Reparatur versucht werden soll
 * @returns {any} Geparste Daten
 */
function parse_json(raw_input, repair = true) {
  try {
    return JSON.parse(raw_input);
  } catch (e) {
    if (repair) {
      // Versuche einfache Reparaturen
      let repaired = raw_input.trim();
      
      // Entferne trailing commas
      repaired = repaired.replace(/,\s*([\]}])/g, '$1');
      
      // Ersetze single quotes mit double quotes (nur bei keys und strings)
      // Dies ist eine vereinfachte Reparatur - in der Praxis komplexer
      try {
        return JSON.parse(repaired);
      } catch (e2) {
        throw new JSONProcessingError(`Invalid JSON: ${e.message}`);
      }
    } else {
      throw new JSONProcessingError(`Invalid JSON: ${e.message}`);
    }
  }
}

/**
 * Lädt ein JSON Schema aus verschiedenen Quellen.
 * 
 * @param {string|Object} schema_source - Pfad zur Schema-Datei oder Schema-Dict oder JSON String
 * @returns {Object} Schema als Dictionary
 */
function load_schema(schema_source) {
  /**
   * Lädt ein JSON Schema aus verschiedenen Quellen.
   * 
   * Args:
   *     schema_source: Pfad zur Schema-Datei oder Schema-Dict oder JSON String
   * 
   * Returns:
   *     Schema als Dictionary
   */
  if (typeof schema_source === 'object' && schema_source !== null) {
    return schema_source;
  }

  // Prüfe ob es ein Pfad ist
  if (typeof schema_source === 'string') {
    try {
      if (fs.existsSync(schema_source)) {
        const content = fs.readFileSync(schema_source, 'utf8');
        return JSON.parse(content);
      }
    } catch (e) {
      if (e instanceof Error && e.name === 'SyntaxError') {
        throw new SchemaValidationError(`Invalid JSON in schema file: ${e.message}`);
      }
      // Falls Datei nicht existiert, weiter mit JSON parsen
    }

    // Versuche als JSON String zu parsen
    try {
      return JSON.parse(schema_source);
    } catch (e) {
      throw new SchemaValidationError(`Schema not found or invalid: ${schema_source}`);
    }
  }

  throw new SchemaValidationError(`Schema not found or invalid: ${schema_source}`);
}

/**
 * Validiert Daten gegen ein JSON Schema.
 * 
 * @param {any} data - Zu validierende Daten
 * @param {string|Object} schema - JSON Schema (Pfad, String oder Dict)
 * @param {string} draft - JSON Schema Draft Version ("auto", "draft7", "2020-12")
 * @returns {boolean} True wenn valid
 */
function validate_with_jsonschema(data, schema, draft = "auto") {
  /**
   * Validiert Daten gegen ein JSON Schema.
   * 
   * Args:
   *     data: Zu validierende Daten
   *     schema: JSON Schema (Pfad, String oder Dict)
   *     draft: JSON Schema Draft Version ("auto", "draft7", "2020-12")
   * 
   * Returns:
   *     True wenn valid
   * 
   * Raises:
   *     SchemaValidationError: Wenn Validierung fehlschlägt
   */
  if (!HAS_JSONSCHEMA && !HAS_AJV) {
    throw new SchemaValidationError("Neither jsonschema nor ajv installed. Run: npm install jsonschema or npm install ajv ajv-formats");
  }

  const schema_dict = load_schema(schema);

  try {
    if (HAS_JSONSCHEMA) {
      const result = validateFunction(data, schema_dict);
      if (!result.valid) {
        const error = result.errors[0];
        const path = error.path || [];
        throw new SchemaValidationError(`Schema validation failed: ${error.message} at ${JSON.stringify(path)}`);
      }
      return true;
    } else if (HAS_AJV) {
      const validate = ajvInstance.compile(schema_dict);
      const valid = validate(data);
      if (!valid) {
        const error = validate.errors[0];
        const path = error.instancePath ? error.instancePath.split('/').filter(p => p) : [];
        throw new SchemaValidationError(`Schema validation failed: ${error.message} at ${JSON.stringify(path)}`);
      }
      return true;
    }
  } catch (e) {
    if (e instanceof SchemaValidationError) {
      throw e;
    }
    throw new SchemaValidationError(`Schema validation error: ${e.message}`);
  }
}

/**
 * Parst, repariert und validiert JSON gegen Schema.
 * 
 * @param {string} raw_input - JSON-String
 * @param {string|Object} schema - JSON Schema
 * @param {boolean} repair - Ob Reparatur versucht werden soll
 * @returns {any} Validierte Daten
 */
function validate_and_convert(raw_input, schema, repair = true) {
  /**
   * Parst, repariert und validiert JSON gegen Schema.
   * 
   * Args:
   *     raw_input: JSON-String
   *     schema: JSON Schema
   *     repair: Ob Reparatur versucht werden soll
   * 
   * Returns:
   *     Validierte Daten
   */
  const data = parse_json(raw_input, repair);
  validate_with_jsonschema(data, schema);
  return data;
}

/**
 * Hilfsklasse zum Erstellen von JSON Schemas.
 */
class SchemaBuilder {
  /** Hilfsklasse zum Erstellen von JSON Schemas. */

  /**
   * Erstellt ein Object-Schema.
   * @param {Object} properties - Eigenschaften
   * @param {string[]} required - Erforderliche Felder
   * @returns {Object} Schema
   */
  static object(properties, required = null) {
    /** Erstellt ein Object-Schema. */
    const schema = {
      type: "object",
      properties: properties
    };
    if (required) {
      schema.required = required;
    }
    return schema;
  }

  /**
   * Erstellt ein String-Schema.
   * @param {string[]} enum_values - Enum-Werte
   * @param {string} pattern - Regex-Pattern
   * @param {number} min_length - Minimale Länge
   * @returns {Object} Schema
   */
  static string(enum_values = null, pattern = null, min_length = null) {
    /** Erstellt ein String-Schema. */
    const schema = {type: "string"};
    if (enum_values) {
      schema.enum = enum_values;
    }
    if (pattern) {
      schema.pattern = pattern;
    }
    if (min_length !== null) {
      schema.minLength = min_length;
    }
    return schema;
  }

  /**
   * Erstellt ein Integer-Schema.
   * @param {number} minimum - Minimalwert
   * @param {number} maximum - Maximalwert
   * @returns {Object} Schema
   */
  static integer(minimum = null, maximum = null) {
    /** Erstellt ein Integer-Schema. */
    const schema = {type: "integer"};
    if (minimum !== null) {
      schema.minimum = minimum;
    }
    if (maximum !== null) {
      schema.maximum = maximum;
    }
    return schema;
  }

  /**
   * Erstellt ein Array-Schema.
   * @param {Object} items - Item-Schema
   * @param {number} min_items - Minimale Anzahl Items
   * @returns {Object} Schema
   */
  static array(items, min_items = null) {
    /** Erstellt ein Array-Schema. */
    const schema = {type: "array", items: items};
    if (min_items !== null) {
      schema.minItems = min_items;
    }
    return schema;
  }
}

// Command line interface
async function main() {
  const { Command } = require('commander');
  const program = new Command();

  program
    .description('JSON Schema Validator')
    .argument('<input>', 'JSON file or string')
    .option('-s, --schema <schema>', 'Schema file')
    .option('-f, --file', 'Input is file')
    .option('-r, --repair', 'Attempt to repair JSON', true)
    .action((input, options) => {
      // Lade Input (Auto-detect file vs string)
      let raw_input;
      if (options.file || (fs.existsSync(input) && fs.statSync(input).isFile())) {
        raw_input = fs.readFileSync(input, 'utf8');
      } else {
        raw_input = input;
      }

      try {
        const result = validate_and_convert(raw_input, options.schema, options.repair);
        console.log(JSON.stringify(result, null, 2));
        console.error("\n✓ Validation passed");
      } catch (e) {
        if (e instanceof JSONProcessingError || e instanceof SchemaValidationError) {
          console.error(`✗ Validation failed: ${e.message}`);
          process.exit(1);
        } else {
          console.error(`✗ Unexpected error: ${e.message}`);
          process.exit(1);
        }
      }
    });

  await program.parseAsync(process.argv);
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = {
  SchemaValidationError,
  load_schema,
  validate_with_jsonschema,
  validate_and_convert,
  SchemaBuilder,
  parse_json,
  JSONProcessingError,
  JSONValidationError
};

File: perl5/json_schema_validator.pl
#!/usr/bin/env perl
# json_schema_validator.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/json_schema_validator.py
# auch in: OpenClaw@gateway2:skills/json-utils/scripts/json_schema_validator.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use JSON::PP qw(decode_json encode_json);
use File::Spec::Functions qw(catfile);
use Getopt::Long;
use Pod::Usage;

# JSON Schema Validator - Validiert JSON gegen JSON Schema Draft 7/2020-12.
# Erweitert Pydantic mit externen Schema-Dateien.

# Globale Variablen für Module
my $HAS_JSON_VALIDATOR;
my $HAS_JSON_XS;

# Versuche JSON::Validator zu laden
eval {
    require JSON::Validator;
    $HAS_JSON_VALIDATOR = 1;
};
if ($@) {
    $HAS_JSON_VALIDATOR = 0;
}

# Versuche JSON::XS zu laden
eval {
    require JSON::XS;
    $HAS_JSON_XS = 1;
};

# JSON Processor Subroutinen (ersetzen json_processor.py)
sub parse_json {
    my ($raw_input, $repair) = @_;
    $repair //= 1;
    
    eval {
        return decode_json($raw_input);
    };
    if ($@ && $repair) {
        # Versuche einfache Reparaturen
        my $repaired = $raw_input;
        $repaired =~ s/,\s*}/}/g;  # Entferne trailing commas
        $repaired =~ s/,\s*\]/]/g;
        eval {
            return decode_json($repaired);
        };
        if ($@) {
            die "JSON processing error: $@";
        }
    } elsif ($@) {
        die "JSON processing error: $@";
    }
}

# Exception Klassen
package SchemaValidationError {
    sub new {
        my ($class, $message) = @_;
        return bless { message => $message }, $class;
    }
    
    sub message {
        my $self = shift;
        return $self->{message};
    }
}

package JSONProcessingError {
    sub new {
        my ($class, $message) = @_;
        return bless { message => $message }, $class;
    }
    
    sub message {
        my $self = shift;
        return $self->{message};
    }
}

# Hauptpaket
package main;

# Hilfsfunktionen
sub load_schema {
    my ($schema_source) = @_;
    
    # Wenn es bereits ein Hash ist
    if (ref($schema_source) eq 'HASH') {
        return $schema_source;
    }
    
    # Prüfe ob es eine Datei ist
    if (-f $schema_source) {
        open my $fh, '<:encoding(UTF-8)', $schema_source or die "Cannot open schema file '$schema_source': $!";
        my $content = do { local $/; <$fh> };
        close $fh;
        
        eval {
            return decode_json($content);
        };
        if ($@) {
            die SchemaValidationError->new("Invalid JSON in schema file: $@");
        }
    }
    
    # Versuche als JSON String zu parsen
    eval {
        return decode_json($schema_source);
    };
    if ($@) {
        die SchemaValidationError->new("Schema not found or invalid: $schema_source");
    }
}

sub validate_with_jsonschema {
    my ($data, $schema, $draft) = @_;
    $draft //= "auto";
    
    if (!$HAS_JSON_VALIDATOR) {
        die SchemaValidationError->new("JSON::Validator not installed. Run: cpan JSON::Validator");
    }
    
    my $schema_dict = load_schema($schema);
    
    my $jv = JSON::Validator->new;
    if ($draft eq "draft7") {
        $jv->schema("http://json-schema.org/draft-07/schema#");
    } elsif ($draft eq "2020-12") {
        # Draft 2020-12 Unterstützung hängt von der Bibliothek ab
        $jv->schema("https://json-schema.org/draft/2020-12/schema");
    }
    
    my @errors = $jv->validate($data, $schema_dict);
    
    if (@errors) {
        my $error_msg = join(", ", map { $_->{message} } @errors);
        die SchemaValidationError->new("Schema validation failed: $error_msg");
    }
    
    return 1;
}

sub validate_and_convert {
    my ($raw_input, $schema, $repair) = @_;
    $repair //= 1;
    
    my $data;
    eval {
        $data = parse_json($raw_input, $repair);
    };
    if ($@) {
        die JSONProcessingError->new($@);
    }
    
    eval {
        validate_with_jsonschema($data, $schema);
    };
    if ($@) {
        die $@;
    }
    
    return $data;
}

# SchemaBuilder Klasse
package SchemaBuilder {
    sub object {
        my ($class, $properties, $required) = @_;
        my $schema = {
            type => "object",
            properties => $properties
        };
        if ($required) {
            $schema->{required} = $required;
        }
        return $schema;
    }
    
    sub string {
        my ($class, $enum, $pattern, $min_length) = @_;
        my $schema = { type => "string" };
        if ($enum) {
            $schema->{enum} = $enum;
        }
        if ($pattern) {
            $schema->{pattern} = $pattern;
        }
        if (defined $min_length) {
            $schema->{"minLength"} = $min_length;
        }
        return $schema;
    }
    
    sub integer {
        my ($class, $minimum, $maximum) = @_;
        my $schema = { type => "integer" };
        if (defined $minimum) {
            $schema->{minimum} = $minimum;
        }
        if (defined $maximum) {
            $schema->{maximum} = $maximum;
        }
        return $schema;
    }
    
    sub array {
        my ($class, $items, $min_items) = @_;
        my $schema = { 
            type => "array", 
            items => $items 
        };
        if (defined $min_items) {
            $schema->{minItems} = $min_items;
        }
        return $schema;
    }
}

# Main Funktion
sub main {
    my ($input, $schema, $file, $repair);
    GetOptions(
        "schema|s=s" => \$schema,
        "file|f"     => \$file,
        "repair|r"   => \$repair,
        "help|h"     => sub { pod2usage(1) }
    ) or pod2usage(2);
    
    if (!@ARGV) {
        pod2usage(2);
    }
    
    $input = $ARGV[0];
    $repair //= 1;
    
    if (!$schema) {
        print STDERR "Error: --schema is required\n";
        exit 1;
    }
    
    # Lade Input (Auto-detect file vs string)
    my $raw_input;
    if ($file || (-f $input)) {
        open my $fh, '<:encoding(UTF-8)', $input or die "Cannot open input file '$input': $!";
        $raw_input = do { local $/; <$fh> };
        close $fh;
    } else {
        $raw_input = $input;
    }
    
    eval {
        my $result = validate_and_convert($raw_input, $schema, $repair);
        print encode_json($result) . "\n";
        print STDERR "\n✓ Validation passed\n";
    };
    if ($@) {
        if ($@->isa('SchemaValidationError') || $@->isa('JSONProcessingError')) {
            print STDERR "✗ Validation failed: " . $@->message . "\n";
        } else {
            print STDERR "✗ Validation failed: $@\n";
        }
        exit 1;
    }
}

# Dokumentation
__END__

=head1 NAME

json_schema_validator.pl - JSON Schema Validator

=head1 SYNOPSIS

json_schema_validator.pl [options] <input>

 Options:
   --schema|-s    Schema file (required)
   --file|-f      Input is file
   --repair|-r    Repair JSON (default)
   --help|-h      Show this help

=head1 DESCRIPTION

Validiert JSON gegen JSON Schema Draft 7/2020-12.

=cut

main() unless caller;
